pkg/kubectl/cmd/top.go

```diff
@@ -17,192 +17,45 @@ limitations under the License.
-func (o *TopOptions) Complete(f *cmdutil.Factory, cmd *cobra.Command, args []string, out io.Writer) error {
-	var err error
-	switch len(args) {
-	case 0:
-		return cmdutil.UsageError(cmd, "resource name is required")
-	case 1:
-		o.ResourceType, err = GetResourceKind(args[0])
-		if err != nil {
-			return err
-		}
-	case 2:
-		o.ResourceType, err = GetResourceKind(args[0])
-		if err != nil {
-			return err
-		}
-		o.ResourceName = args[1]
-	default:
-		return cmdutil.UsageError(cmd, cmd.Use)
-	}
-
-	if o.ResourceType == api.Kind("Node") && o.PrintContainers {
-		return cmdutil.UsageError(cmd, "flag --containers is not allowed for nodes")
-	}
-
-	o.Namespace, _, err = f.DefaultNamespace()
-	if err != nil {
-		return err
-	}
-	cli, err := f.Client()
-	if err != nil {
-		return err
-	}
-	o.Client = metricsutil.DefaultHeapsterMetricsClient(cli)
-	o.Printer = metricsutil.NewTopCmdPrinter(out)
-	return nil
-}
 
-// Validate validates all the required options for top.
-func (o TopOptions) Validate() error {
-	err := validateHandledResourceType(o.ResourceType)
-	if err != nil {
-		return err
-	}
-	return nil
+	// create subcommands
+	cmd.AddCommand(NewCmdTopNode(f, out))
+	cmd.AddCommand(NewCmdTopPod(f, out))
+	return cmd
 }
 
 // RunTop implements all the necessary functionality for top.
-func (o TopOptions) RunTop() error {
-	switch o.ResourceType {
-	case api.Kind("Node"):
-		metrics, err := o.Client.GetNodeMetrics(o.ResourceName, o.Selector)
-		if err != nil {
-			return err
-		}
-		return o.Printer.PrintNodeMetrics(metrics)
-	case api.Kind("Pod"):
-		metrics, err := o.Client.GetPodMetrics(o.Namespace, o.ResourceName, o.AllNamespaces, o.Selector)
-		if err != nil {
-			return err
-		}
-		return o.Printer.PrintPodMetrics(metrics, o.PrintContainers)
-	}
-	return errors.New("unsupported resource requested")
-}
-
-func GetResourceKind(resourceType string) (unversioned.GroupKind, error) {
-	if kind, found := AliasesToKind[resourceType]; found {
-		return kind, nil
-	}
-	return unversioned.GroupKind{}, errors.New("unsupported resource requested")
-}
-
-var AliasesToKind map[string]unversioned.GroupKind = GetAliasesToKind(HandledResources)
-
-func GetAliasesToKind(resources []unversioned.GroupKind) map[string]unversioned.GroupKind {
-	aliasesToKind := make(map[string]unversioned.GroupKind, 0)
-	for _, res := range resources {
-		resName := strings.ToLower(res.Kind)
-		aliasesToKind[resName] = res
-		for _, alias := range kubectl.ResourceAliases([]string{resName}) {
-			aliasesToKind[alias] = res
-		}
-	}
-	return aliasesToKind
-}
-
-func validateHandledResourceType(resourceType unversioned.GroupKind) error {
-	for _, r := range HandledResources {
-		if r == resourceType {
-			return nil
-		}
-	}
-	return errors.New("unsupported resource requested")
+func (o TopOptions) RunTop(f *cmdutil.Factory, cmd *cobra.Command, args []string, out io.Writer) error {
+	return cmdutil.UsageError(cmd, "Required resource not specified.")
```

Comment:
It should print out help.  Config does this already.

On Aug 8, 2016, at 1:22 PM, David Eads <notifications@github.com> wrote:

In pkg/kubectl/cmd/top.go
<https://github.com/kubernetes/kubernetes/pull/28844#discussion_r73915874>:

> -		for _, alias := range kubectl.ResourceAliases([]string{resName}) {
> -			aliasesToKind[alias] = res
> -		}
> -	}
> -	return aliasesToKind
> -}
> -
> -func validateHandledResourceType(resourceType unversioned.GroupKind) error {
> -	for _, r := range HandledResources {
> -		if r == resourceType {
> -			return nil
> -		}
> -	}
> -	return errors.New("unsupported resource requested")
> +func (o TopOptions) RunTop(f *cmdutil.Factory, cmd *cobra.Command, args []string, out io.Writer) error {
> +	return cmdutil.UsageError(cmd, "Required resource not specified.")

I'm not yet sure what the bare 'top' command should do if we have the
subcommands. Any suggestions on that?

@fabianofranz <https://github.com/fabianofranz> got a link to our generic
"empty subcommand" command?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/28844/files/71e93b2a737469aa3800aede4018488dc80e3da9..af5b468f0e8a39826664f67790513d45bdd067e0#r73915874>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p_fk3exITwwEcjc-0gYngkLV1D9Kks5qd2W7gaJpZM4JKfgH>
.